hub.go

```diff
@@ -208,75 +247,234 @@ func checkCmd(m []byte) {
 				behaviour = args[4]
 			}
 
-			err := Tools.Download(pack, tool, toolVersion, behaviour)
+			reportPendingProgress := func(msg string) {
+				mapD := map[string]string{"DownloadStatus": "Pending", "Msg": msg}
+				mapB, _ := json.Marshal(mapD)
+				hub.broadcastSys <- mapB
+			}
+			err := hub.tools.Download(pack, tool, toolVersion, behaviour, reportPendingProgress)
 			if err != nil {
 				mapD := map[string]string{"DownloadStatus": "Error", "Msg": err.Error()}
 				mapB, _ := json.Marshal(mapD)
-				h.broadcastSys <- mapB
+				hub.broadcastSys <- mapB
 			} else {
 				mapD := map[string]string{"DownloadStatus": "Success", "Msg": "Map Updated"}
 				mapB, _ := json.Marshal(mapD)
-				h.broadcastSys <- mapB
+				hub.broadcastSys <- mapB
 			}
 		}()
 	} else if strings.HasPrefix(sl, "log") {
-		go logAction(sl)
+		go hub.logAction(sl)
 	} else if strings.HasPrefix(sl, "restart") {
+		// potentially, the sysStray dependencies can be removed  https://github.com/arduino/arduino-create-agent/issues/1013
 		log.Println("Received restart from the daemon. Why? Boh")
-		Systray.Restart()
+		hub.systray.Restart()
 	} else if strings.HasPrefix(sl, "exit") {
-		Systray.Quit()
+		hub.systray.Quit()
 	} else if strings.HasPrefix(sl, "memstats") {
-		memoryStats()
+		hub.memoryStats()
 	} else if strings.HasPrefix(sl, "gc") {
-		garbageCollection()
+		hub.garbageCollection()
 	} else if strings.HasPrefix(sl, "hostname") {
-		getHostname()
+		hub.getHostname()
 	} else if strings.HasPrefix(sl, "version") {
-		getVersion()
+		hub.getVersion()
 	} else {
-		go spErr("Could not understand command.")
+		go hub.spErr("Could not understand command.")
 	}
 }
 
-func logAction(sl string) {
+type logWriter struct {
+	onWrite func([]byte)
+}
```

Comment:
I would create an utility that transform a channel to a writer, for instance something like that
```go
// ChanWriter is a simple io.Writer that sends data to a channel.
type ChanWriter struct {
	Ch chan<- []byte
}

func (u *ChanWriter) Write(p []byte) (n int, err error) {
	u.Ch <- p
	return len(p), nil
}
```
then you can directly use it as follows
```go
multiWriter := io.MultiWriter(&ChanWriter{Ch: hub.broadcastSys}, os.Stderr)
```